### ISSUE TITLE
Support for malloc's from inside libc (eg. strdup or fopen)

### ISSUE THREAD

**Issue by wapiflapi** [wapiflapi/villoc#12]
Right now the easiest way of using villoc is with `ltrace` as documented in the README. This works fine in some cases but fails spectacularly when libc it's doing its own internal allocations, and especially if those are later freed from the target. (eg `free(strdup())`.

Right now the best (only?) idea I have is shipping a gdb script that sets some breakpoints and outputs something compatible with villoc's input.

If anyone has a better idea I'd be happy to talk about it!

**Comment by wapiflapi** [comment]
Okay quick and dirty gdb script `villoc.gdb`:

```gdb
set logging off
set breakpoint pending on

define on_malloc
  set $size = $rdi
  finish
  printf "@VILLOC@ malloc(%d) = %p\n", $size, $rax
  continue
end

define on_calloc
  set $nmemb = $rdi
  set $size = $rsi
  finish
  printf "@VILLOC@ calloc(%d, %d) = %p\n", $nmemb, $size, $rax
  continue
end

define on_realloc
  set $ptr = $rdi
  set $size = $rsi
  finish
  printf "@VILLOC@ realloc(%p, %d) = %p\n", $ptr, $size, $rax
  continue
end

define on_free
  set $ptr = $rdi
  finish
  printf "@VILLOC@ free(%p) = %p\n", $ptr, $rax
  continue
end

break __GI___libc_malloc
commands
  on_malloc
end

break __GI___libc_calloc
commands
  on_calloc
end

break __GI___libc_realloc
commands
  on_realloc
end

break __GI___libc_free
commands
  on_free
end

run

```

And then to invoke it something like:

```bash
gdb --quiet --batch -command=./exploit/villoc.gdb \
    --args $@ |& tee gdb.logs

grep "@VILLOC@" gdb.logs | sed 's/@VILLOC@ //' > gdbtrace
villoc.py gdbtrace villoc.html
```

This seems to be a bit better in some cases, but I still have times where blocks are freed out of nowhere in legit software.


edit: Ps: the `define on_X` is a workaround to be able to exec a command after `finish`, I don't know how to do this easier.

**Comment by ampotos** [comment]
You can do that with using a DBI framework instead of gdb. It would be faster and, depending of the framework, may not need have any dependencies.

**Comment by ZetaTwo** [comment]
This could also be done with LD_PRELOAD hooks, right? At least for dynamically linked executables.

**Comment by wapiflapi** [comment]
The problem with LD_PRELOAD is it doesn't track calls made by libc itself,
for example in `strdup()`, right? At least I don't think it would.

**Comment by ZetaTwo** [comment]
I might be missing something but I really think it looks like this is hooking the malloc inside the strdup.

```
> cat mallocprint.c
#define _GNU_SOURCE

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <dlfcn.h>

void *(*original_malloc)(size_t);
char *(*original_strdup)(const char *);
void (*original_free)(void *);
__attribute__((constructor)) void test_malloc()
{
    puts("start\n");
	original_malloc = dlsym(RTLD_NEXT, "malloc");
    original_strdup = dlsym(RTLD_NEXT, "strdup");
    original_free = dlsym(RTLD_NEXT, "free");
}

char *strdup(const char *s) {
    puts("strdup call");
    char* s2 = original_strdup(s);
    puts("strdup ret");
    return s2;
}


void* malloc (size_t size) {
    char *msg  = "malloc call\n";
    char *msg2 =  "malloc ret\n";
    write(1, msg, 14);
    original_malloc = dlsym(RTLD_NEXT, "malloc");
    void* res = original_malloc(size);
    write(1, msg2, 13);
    return res;
}

void free (void *s) {
    puts("free call");
    original_free(s);
    puts("free ret");
    return;
}
```
```
> cat test.c 
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

const char *delim = "---\n";

int main() {
    char *str1 = "test1\n";

    write(2, delim, 5);
    char *str2 = strdup(str1);
    write(2, delim, 5);
    char *str3 = strdup(str1);
    char *str4 = strdup(str1);
    write(2, delim, 5);


    printf("%p: %s", str1, str1);
    printf("%p: %s", str2, str2);
    printf("%p: %s", str3, str3);
    printf("%p: %s", str4, str4);

    free(str2);
    free(str3);
    free(str4);

    return 0;
}
```

```
> gcc -Wall -o printmalloc.so mallocprint.c -shared -fPIC -ldl
> gcc -Wall -o test test.c
> LD_PRELOAD=./printmalloc.so ./test
malloc call
mmalloc ret
fstart

---
strdup call
malloc call
mmalloc ret
fstrdup ret
---
strdup call
malloc call
mmalloc ret
fstrdup ret
strdup call
malloc call
mmalloc ret
fstrdup ret
---
0x55f06f879909: test1
0x55f070017670: test1
0x55f070017690: test1
0x55f0700176b0: test1
free call
free ret
free call
free ret
free call
free ret
```

**Comment by trpyn** [comment]
I think by hooking strdup here, you will have duplicate calls to malloc since strdup calls malloc.

We just need to hook *alloc, I've re-tested the pull request #5 with your code calling strdup and it seems to work fine to catch the allocs, but misses some free.

This is the output from valgrind, stating 4 allocs and 4 frees.

```
$ valgrind ./a.out 
==5748== Memcheck, a memory error detector
==5748== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==5748== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==5748== Command: ./a.out
==5748== 
---
---
---
0x10a009: test1
0x4a42040: test1
0x4a42090: test1
0x4a420e0: test1
==5748== 
==5748== HEAP SUMMARY:
==5748==     in use at exit: 0 bytes in 0 blocks
==5748==   total heap usage: 4 allocs, 4 frees, 1,045 bytes allocated
==5748== 
==5748== All heap blocks were freed -- no leaks are possible
==5748== 
==5748== For counts of detected and suppressed errors, rerun with: -v
==5748== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

And the output using my implem:
```
$ cat output 
malloc(7) = 0x162104cad40
malloc(7) = 0x162104cad60
malloc(7) = 0x162104cad80
malloc(1024) = 0x162104cada0
free(0x162104cad40) = <void>
free(0x162104cad60) = <void>
free(0x162104cad80) = <void>
```
So 4 allocs, but only 3 frees.
I believe the 4th alloc and free are made by printf when buffering the line. However I still don't know why the 4th free() isn't reported.

**Comment by wapiflapi** [comment]
From what I understand the hook to `strdup` was only for demonstration and
print purposes, right?
I haven't looked in to this yet but I plan on doing so today.

**Comment by ZetaTwo** [comment]
Yes, that's only to more clearly show what's going on. It doesn't affect the number of allocations.

**Comment by wapiflapi** [comment]
So looking into this, it indeed seems to work. Thanks a lot for the PoC @ZetaTwo, I was under the **wrong** impression that this wouldn't work.

With a test program doing:
```
wapiflapi@box:~/Projects/villoc/tests$ ltrace ./a.out
malloc(1024)                                                                        = 0x55584e2a4260
strdup("some string")                                                               = 0x55584e2a4670
fopen("/dev/null", "r")                                                             = 0x55584e2a4690
getcwd(0, 4096)                                                                     = "/home/wapiflapi/Projects/villoc/"...
free(0x55584e2a4260)                                                                = <void>
free(0x55584e2a4670)                                                                = <void>
realloc(0x55584e2a4690, 0)                                                          = 0
realloc(0x55584e2a48c0, 0)                                                          = 0
+++ exited (status 0) +++
```

Using `ltrace`:
```
wapiflapi@box:~/Projects/villoc/tests$ ltrace ./a.out  |& ../villoc.py - out.html
malloc(1024)                                     = 0x55a1702a2260
free(0x55a1702a2260)                             = <void>
free(0x55a1702a2670)                             = <void>
realloc(0x55a1702a2690, 0)                       = 0
realloc(0x55a1702a28c0, 0)                       = 0
ignoring line: +++ exited (status 0) +++

```
![screenshot from 2019-01-21 10-03-04](https://user-images.githubusercontent.com/1619783/51463592-dee49180-1d63-11e9-821e-828a323b56ce.png)

Now, using `LD_PRELOAD`:
```
wapiflapi@box:~/Projects/villoc/tests$ LD_PRELOAD=./printmalloc.so ./a.out  |& ../villoc.py - out.html
malloc(1024) = 0x560bf6146260
malloc(12) = 0x560bf6146670
malloc(552) = 0x560bf6146690
malloc(4096) = 0x560bf61468c0
free(0x560bf6146260) = <void>
free(0x560bf6146670) = <void>
realloc(0x560bf6146690, 0) = (nil)
realloc(0x560bf61468c0, 0) = (nil)
```
![screenshot from 2019-01-21 10-03-16](https://user-images.githubusercontent.com/1619783/51463584-d8561a00-1d63-11e9-89f3-08030bee81f9.png)

# How to add this to the project?

I'm going to do some more tests, especially with the usecase I had that initially prompted this issue then I'll probably add this way of doing things to `master`. What do y 'all think is a good way of doing this?

I was thinking about a `tracer` folder where we could have a `ld_preload` sub folder with this ? Then we could add things like #13 in other `tracer` sub-folders ?

**Comment by wapiflapi** [comment]
Okay not trivial.

https://github.com/jtolds/malloc_instrumentation/blob/9f8387bb270dd194e57aca38dc45039f4e54cc0a/malloc_instrument.c

```
// Unfortunately, it's not quite as straightforward as it sounds, as fprintf
// and dlsym both use heap-based memory allocation. During initialization, we
// use a dummy implementation of malloc that uses a small buffer. Once dlsym
// loading is done, then we switch to our real implementation, which, unless
// a recursive mutex is already held, first outputs the call arguments, makes
// the call, and then outputs the return value. If the recursive mutex is
// already held, then the call was due to some call made while outputting
// arguments, so we just forward the call along to the real call.
```
We might want to use that library, except we don't want it to call malloc during the printing either (because logging shouldn't mess with the heap while figuring out an exploit) but that's do-able.

**Comment by wapiflapi** [comment]
#13 is a solution to this problem, I'll update the README soon and then close this.

**Comment by wapiflapi** [comment]
Fixed by #13
